### PR TITLE
doc/website/blog: usability blog

### PR DESCRIPTION
**Pull request by rotemtam**

### CODE REVIEW

File: doc/website/blog/2024-04-04-top-usability-issues.mdx [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+
+### Building a loveable migration tool
+
+> **"I just love dealing with migrations!"**
+>
+> -No developer, ever.
+
+Over the past three years, [Ariel](https://github.com/a8m), my co-founder and I (along with the rest of our team at
+[Ariga](https://ariga.io)), have been working on [Atlas](https://atlasgo.io), a database schema-as-code tool.
+After many years of building software professionally, we've come to realize that the most stressful, tedious and error-prone part of
```
Comment (a8m) [inline doc/website/blog/2024-04-04-top-usability-issues.mdx:18]:
```suggestion
After many years of building software professionally, we've come to realize that one of the most stressful, tedious and error-prone part of
```

File: doc/website/blog/2024-04-04-top-usability-issues.mdx [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+current state of the database is, then they need to know what the desired state is, and then they need to write
+a script that will take the database from the current state to the desired state.  Some changes are trivial, but
+some require consideration and research.
+
+**Solution**: Use a tool that can [automatically generate migration](/versioned/diff) scripts for you.
+This way, you can focus on the desired state of the database and let the tool figure out how to get there.
+
+:::
+
+### Issue #3: Working in parallel
```
Comment (a8m) [inline doc/website/blog/2024-04-04-top-usability-issues.mdx:104]: Maybe you can link to [this](https://atlasgo.io/lint/analyzers#non-linear-changes) in the problem/solution section.

**Review by a8m — APPROVED** [review]
🚀

File: doc/website/blog/2024-04-04-top-usability-issues.mdx [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+working on different features that require different changes to the database schema.  This can lead to conflicts
+when developers try to merge their changes together.
+
+Classic migration tools don't provide a good way to detect and handle this situation.  Because each migration
+script is created in its own file, common conflict detection tools like git can't help you.  Your source control
+system can't tell you if two developers are working on the same table, or if they are adding columns with the same
+name.
+
+For this reason it's common to see teams surprised in production when conflicting migrations are applied.
+Even worse, in some cases, migrations may be applied out of order or skipped entirely, leading to an
```
Comment (hilakashai) [inline doc/website/blog/2024-04-04-top-usability-issues.mdx:116]:
```suggestion
Even worse, in some cases migrations may be applied out of order or skipped entirely, leading to an
```

File: doc/website/blog/2024-04-04-top-usability-issues.mdx [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+inconsistent and unknown state of the database.
+
+:::info Issue #3: Working in parallel
+
+**Problem**: Source control systems can't help you detect conflicts between migration scripts.  This can lead to
+undetected issues in production when they are deployed. To overcome this, team develop their own processes to
```
Comment (hilakashai) [inline doc/website/blog/2024-04-04-top-usability-issues.mdx:122]:
```suggestion
undetected issues in production when they are deployed. To overcome this, teams develop their own processes to
```

File: doc/website/blog/2024-04-04-top-usability-issues.mdx [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+### Issue #4: Tracking partial failures
+
+Virtually all classic migration tools maintain a metadata table on the target database that tracks which migrations
+have been applied. Using this metadata, the tool can determine which migrations need to be applied to bring the
+database up to date.
+
+When migrations succeed, everything is great. But in my experience, when migrations fail, especially when they fail
+partially, it can be a nightmare to recover from.
+
+Suppose you planning a migration to version `N+1` of the database, from version `N`.  The migration script for version
```
Comment (hilakashai) [inline doc/website/blog/2024-04-04-top-usability-issues.mdx:139]:
```suggestion
Suppose you planning a migration to version `N+1` of the database, from version `N`. The migration script for version
```

File: doc/website/blog/2024-04-04-top-usability-issues.mdx [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+`N` contains 10 changes.  The migration fails after applying the 5th change.  If you use a database that supports fully
```
Comment (hilakashai) [inline doc/website/blog/2024-04-04-top-usability-issues.mdx:140]:
```suggestion
`N` contains 10 changes. The migration fails after applying the 5th change. If you use a database that supports fully
```

File: doc/website/blog/2024-04-04-top-usability-issues.mdx [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+transactional DDL (like Postgres), and all of your changes are transactional, then you are in luck - your migration tool
+can safely roll back the changes that were applied and your database (revisions table) will remain in version `N1`.
+
+But if you are using a database that doesn't support transactional DDL, or if your changes are not transactional, then
+you are in trouble.  The migration tool can't rollback the changes that were applied, and your database will in a state that
```
Comment (hilakashai) [inline doc/website/blog/2024-04-04-top-usability-issues.mdx:145]:
```suggestion
you are in trouble. The migration tool can't rollback the changes that were applied, and your database will in a state that
```

File: doc/website/blog/2024-04-04-top-usability-issues.mdx [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+is somewhere in the middle between version `N1` and version `N+1`.  All of the migration tools that I know of don't
```
Comment (hilakashai) [inline doc/website/blog/2024-04-04-top-usability-issues.mdx:146]:
```suggestion
is somewhere in the middle between version `N1` and version `N+1`. All of the migration tools that I know of don't
```

File: doc/website/blog/2024-04-04-top-usability-issues.mdx [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+> the dirty state persists, which prevents attempts to run more migrations on top of a failed migration.
+
+And the solution?
+
+> After you clean up you database you can also open schema_migrations table and
+> change Dirty flag and rollback version number to last migration that was successfully applied.
+
+:::info Issue #4: Tracking partial failures
+
+**Problem**: Classic migration tools don't handle partial failures well.  When a migration fails, especially when it fails
```
Comment (hilakashai) [inline doc/website/blog/2024-04-04-top-usability-issues.mdx:169]:
```suggestion
**Problem**: Classic migration tools don't handle partial failures well. When a migration fails, especially when it fails
```

File: doc/website/blog/2024-04-04-top-usability-issues.mdx [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+partially, it can leave your database in an unknown state with the revision table out of sync with the actual state of the
+database.  Resolving this issue requires manual handling of the database which is dangerous and error-prone.
```
Comment (hilakashai) [inline doc/website/blog/2024-04-04-top-usability-issues.mdx:171]:
```suggestion
database. Resolving this issue requires manual handling of the database which is dangerous and error-prone.
```

File: doc/website/blog/2024-04-04-top-usability-issues.mdx [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+
+**Solution**: Use a tool that natively supports [transactions](/versioned/apply#transaction-configuration) where possible,
+and uses [statement-level granularity](/blog/2023/04/10/troubleshooting-migrations#statement-level-granularity) to
+keep track of applied changes.
+
+:::
+
+### Issue #5: Pre-planned rollbacks
+
+One of the most curious things about classic migration tools is the prevalence of the `down` migration.  The idea is
```
Comment (hilakashai) [inline doc/website/blog/2024-04-04-top-usability-issues.mdx:181]:
```suggestion
One of the most curious things about classic migration tools is the prevalence of the `down` migration. The idea is
```

File: doc/website/blog/2024-04-04-top-usability-issues.mdx [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+that whenever you write a migration script that to take the database from version `N` to version `N+1`, then you should
```
Comment (hilakashai) [inline doc/website/blog/2024-04-04-top-usability-issues.mdx:182]:
```suggestion
that whenever you write a migration script to take the database from version `N` to version `N+1`, then you should
```

File: doc/website/blog/2024-04-04-top-usability-issues.mdx [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+also write the `down` migration that will take the database from version `N+1` back to version `N`.
+
+Why is this curious? Because after interviewing many developers across many organizations in virtually every industry,
+I have learned that in practice nobody uses `down` migrations, especially in production.
+
+We have recently written [extensively on this topic](/blog/2024/04/01/migrate-down), but the gist is that down migrations
+are never used because:
+
+* *They are naive.* Pre-planned rollbacks assume that all statements in the `up` migration have been applied successfully. In practice,
+  rolling back a version happens when things did not work as planned, making the pre-planned rollback obselete.
```
Comment (hilakashai) [inline doc/website/blog/2024-04-04-top-usability-issues.mdx:192]:
```suggestion
  rolling back a version happens when things did not work as planned, making the pre-planned rollback obsolete.
```

File: doc/website/blog/2024-04-04-top-usability-issues.mdx [hunk 14]
```diff
@@ -0,0 +1,233 @@
...
+  deploying the previous version of the application. When it comes to versions of our application code, this works
+  perfectly. We pull the container image that corresponds to the previous version, and we deploy it.
+
+  But what about the database? When we pull artifacts from a previous version, they do not contain the down
+  files that are needed to revert the database changes back to the necessary schema - they were only created in a future commit!
+
+:::info Issue #5: Pre-planned rollbacks
+
+**Problem**: Pre-planned rollbacks ("down migrations") are never used in practice. They are naive, destructive,
+and incompatible with broader rollback mechanisms.  This leads to a false sense of security which is unraveled
```
Comment (hilakashai) [inline doc/website/blog/2024-04-04-top-usability-issues.mdx:209]:
```suggestion
and incompatible with broader rollback mechanisms. This leads to a false sense of security which is unraveled
```

File: doc/website/blog/2024-04-04-top-usability-issues.mdx [hunk 15]
```diff
@@ -0,0 +1,233 @@
...
+when a rollback is actually needed and realize that they will need to handle it manually.
```
Comment (hilakashai) [inline doc/website/blog/2024-04-04-top-usability-issues.mdx:210]:
```suggestion
when a rollback is actually needed and the user realizes that they will need to handle it manually.
```

File: doc/website/blog/2024-04-04-top-usability-issues.mdx [hunk 16]
```diff
@@ -0,0 +1,233 @@
...
+
+**Solution**: Use a tool that support dynamically-planned, [safe rollbacks](/blog/2024/04/01/migrate-down) that
```
Comment (hilakashai) [inline doc/website/blog/2024-04-04-top-usability-issues.mdx:212]:
```suggestion
**Solution**: Use a tool that supports dynamically-planned, [safe rollbacks](/blog/2024/04/01/migrate-down) that
```

File: doc/website/blog/2024-04-04-top-usability-issues.mdx [hunk 17]
```diff
@@ -0,0 +1,233 @@
...
+can revert the database to a previous state without data loss even in cases of partial failures.
+
+:::
+
+### Try Atlas today
+
+I believe these usability issues should not be taken lightly. Your team's ability to move fast, refactor and
+respond to changing requirements is limited by the agility in which you can evolve your database schema.
+One of the most interesting things that we see with teams that adopt Atlas is that we see them moving from
```
Comment (hilakashai) [inline doc/website/blog/2024-04-04-top-usability-issues.mdx:221]:
```suggestion
One of the most interesting things that we see with teams that adopt Atlas is seeing them move from
```

File: doc/website/blog/2024-04-04-top-usability-issues.mdx [hunk 18]
```diff
@@ -0,0 +1,233 @@
...
+occasional, dreaded, avoided at all costs migrations to planning and deploying hundreds of schema changes
```
Comment (hilakashai) [inline doc/website/blog/2024-04-04-top-usability-issues.mdx:222]:
```suggestion
occasional, dreaded, avoided-at-all-costs migrations, to planning and deploying hundreds of schema changes
```

File: doc/website/blog/2024-04-04-top-usability-issues.mdx [hunk 19]
```diff
@@ -0,0 +1,233 @@
...
+a year. This means that they can move faster, respond to customer feedback more quickly, and innovate more
+effectively.
+
+To try Atlas, head over to the [Getting Started](/getting-started) today!
+
+### Wrapping up
+
+In this article, we have discussed the top 5 usability issues with migration tools today and hinted at how modern
+migration tools (like Atlas) can address them.  I
```
Comment (hilakashai) [inline doc/website/blog/2024-04-04-top-usability-issues.mdx:231]:
```suggestion
migration tools (like Atlas) can address them. 
```

**Pull request merged**